Issue/1175/oas 3 enhance validation and random object generation ts

tschlat opened the pull request:
Here's the third PR, introducing two main features:

- RandomValueGenerator for generating random OpenAPI data.
- OpenAPI Schema Validator leveraging com.atlassian.oai:swagger-request-validator-core for OpenAPI validation.

bbortt commented:
@tschlat build is failing: https://github.com/citrusframework/citrus/actions/runs/13526843489/job/37799461107?pr=1325#step:6:42458.

tschlat commented:
There was a dependency missing. Pipeline is green. Feel free to review.

tschlat commented:
Any activities around this? So that we can move ahead?

bbortt commented:
> Any activities around this? So that we can move ahead?

52/104 files. 27 comments so far 😉

bbortt commented on connectors/citrus-openapi/src/test/java/org/citrusframework/openapi/actions/OpenApiClientActionBuilderTest.java:
```diff
@@ -1,25 +1,23 @@
 package org.citrusframework.openapi.actions;
```
some more tests with missing copyright headers. there's this function in Intellij, where you can apply copyright headers to all files matching a glob pattern (e.g. `**/*.java`). maybe you should try it 😉 or I can do it, if you want me to.

bbortt replied:
@christophd fixed this. thanks!

bbortt reviewed (changes requested):
finally the long awaited review 😉 hope it's all right. let me know if you have any questions or if I got something wrong...

christophd commented on connectors/citrus-openapi/pom.xml:
```diff
@@ -72,6 +80,13 @@
     </dependency>
+    <dependency>
+      <groupId>uk.org.webcompere</groupId>
+      <artifactId>system-stubs-core</artifactId>
```
Okay noticed, if there is more PRs to come that use this library I am fine with it. So to completely understand:  This is a substitute to the `EnvironmentProvider` which is a tooling on your fork, right?

tschlat replied:
Yes that was right. Since this came up again I surrendered an implemented the logic using the existing SystemProvider which is straight forward without any magic.

christophd commented on connectors/citrus-openapi/src/main/resources/META-INF/citrus/message/validator/openApi:
```diff
@@ -0,0 +1,2 @@
+name=defaultOpenApiMessageValidator
+type=org.citrusframework.openapi.validation.OpenApiSchemaValidation
```
During review it felt a bit weird that we have the same class declared as a schema validator and message validator. First I thought this is a copy and paste error, but I believe this is meant to be on purpose, right?

christophd commented on connectors/citrus-openapi/src/test/java/org/citrusframework/openapi/integration/OpenApiClientIT.java:
```diff
@@ -17,83 +17,209 @@
+            .response(OK)
+            .message()
+            .type(MessageType.JSON)
+            .validate(openApi(null)
```
Having to pass `null` as an argument is not very comfortable. Please think of adding an empty argument method for this. Just to be sure that I get it right. Using the `null` argument here would perform a `normal` schema validation without the OpenAPI spec being included?

tschlat replied:
Method was already there, wating for being used.

But no. There is no normal schema validation. For this, we would have to register the individual components of the OpenApi as json schemas - i guess. I don't know if there is a use case for this. The implementation of  OpenApi SchemaValidation follows the pattern of JsonSchemaValidation. Get repository by name and resolve schema from it. Therefore, it is not required to pass in an OpenApiSpecification. If you do so, the default enablement of the context would be derived from that spec. So for the validation to succeed, an OpenApiSpec needs to be registered.

christophd reviewed (approved):
Thanks for your work! In general I am good with the PR.

I have opened a review PR #1332 to incorporate some improvements (e.g. missing license headers)

As a follow-up it would be good to think about what needs to be adjusted in the documentation regarding these changes made here.

bbortt commented:
> Okay noticed, if there is more PRs to come that use this library I am fine with it. So to completely understand: This is a substitute to the EnvironmentProvider which is a tooling on your fork, right?

regarding `system-stubs-core`, the `SystemProvider` (sorry, didn't remember the exact name before) is not something on our fork. it's here: https://github.com/citrusframework/citrus/blob/main/core/citrus-base/src/main/java/org/citrusframework/util/SystemProvider.java. it makes it possible to `@mock` this during tests.

christophd commented:
I am preparing for a new Citrus release 4.6.0
Should this be part of it?

bbortt commented:
I think @tschlat hasn't worked on my comments yet. also he mentioned some things he noticed while reading my comments, that he wanted to take a look at. I would wait with it.

christophd commented:
Okay no problem - there will be other releases 😄

tschlat commented:
Hey guys, i had to replace the apicurio api by the swagger API. This was more or less straight forward but still quite a lot of files to touch. The reason for this was:
1. The API Version was outdated and was not able to read the current swagger version
2. The API did not provide the "required" property for body, such that it was not possible to determine whether to randomly generate a body or not. I need to be able to distinguish here on "required" and "non required".
3. We had two APIs for handling OpenAPI and the Swagger API covers all aspects and is required for Atlassians validation API.

So i went for Swagger API.

I committed the relevant changes and I will now look into the comments in the MR.

tschlat commented:
> Okay no problem - there will be other releases 😄

Sry - I was on vacation.

christophd commented:
> This was more or less straight forward but still quite a lot of files to touch.

I would try to divide this into smaller chunks. Putting even more changes into the PR makes it almost impossible to review

tschlat commented:
Agreed. I didn't really think about this. I will try to postpone this change into the next PR.

tschlat commented on connectors/citrus-openapi/src/main/java/org/citrusframework/openapi/actions/OpenApiClientRequestActionBuilder.java:
```diff
@@ -18,142 +18,292 @@
+
+                body.ifPresent(oasSchema -> {
+
+                    if (autoFill == AutoFillType.ALL || autoFill == AutoFillType.REQUIRED) {
```
No. I did not think of this approach. But now, that you mention this, I have added a method:
`shouldFill` the the enum. This can be used at least at three locations.

tschlat commented on connectors/citrus-openapi/src/main/java/org/citrusframework/openapi/actions/OpenApiPayloadBuilder.java:
```diff
@@ -0,0 +1,54 @@
+        super(payload);
+    }
+
+    private static void replaceDynamicContentInMultiValueMap(TestContext context, MultiValueMap<Object, Object> multiValueMap) {
```
I see the theoretical issue of a concurrent access but then again, the builder is actually not meant for concurrent access. It is used internally by the scenario to actually build the message. All threads that would be manipulating that content during or after message generation already have the potential to run in a concurrent acces without me, finally modifying the content. So concurrent access on the message builder would cause unpredictable results anyways. Therefore I'd stick with the solution unless you come up with an idea how to solve this in a better way.

tschlat commented on connectors/citrus-openapi/src/main/java/org/citrusframework/openapi/actions/OpenApiServerResponseActionBuilder.java:
```diff
@@ -16,118 +16,302 @@
+            openApiSpecification.getOperation(operationId, context)
+                .ifPresentOrElse(operationPathAdapter ->
+                    fillRandomData(openApiSpecification, operationPathAdapter, context), () -> {
+                    throw new CitrusRuntimeException(
```
Here on builder level, it is very probable, that a tester explicitly calls the builder with a valid operationId, which is totally fine. The adapter is able to resolve the operation from the operationId. If no operationId is available on an operation, it would still be valid to pass in a "uniqueOperationId" here.

But i get the point. operationId is missleading here. I have now replaced all occurrences of operationId by operationKey where operationKey could be both the operationId or the operationUniqueId.

tschlat commented on connectors/citrus-openapi/src/main/java/org/citrusframework/openapi/actions/OpenApiServerResponseActionBuilder.java:
```diff
@@ -16,118 +16,302 @@
+                String mediaTypeName = schemaForMediaType.adapted();
+
+                // Support any json for now. Especially: application/json, application/json;charset=UTF-8
+                if (mediaTypeName.toUpperCase().contains("JSON")) {
```
I replaced this with explicit checks against application/json and application/json;charset=UTF-8, the later being deprecated

tschlat commented on connectors/citrus-openapi/src/main/java/org/citrusframework/openapi/model/v2/Oas20ModelHelper.java:
```diff
@@ -103,30 +134,63 @@ public static Map<String, OasSchema> getHeaders(Oas20Response response) {
     }
 
+    private static boolean isFormDataMediaType(String type) {
+        return Arrays.asList(APPLICATION_FORM_URLENCODED_VALUE, MULTIPART_FORM_DATA_VALUE).contains(type);
```
Sure. I replaced this with a static set.

tschlat commented on connectors/citrus-openapi/src/main/java/org/citrusframework/openapi/random/RandomElement.java:
```diff
@@ -0,0 +1,118 @@
+ * Interface representing a random element in a JSON structure. This interface provides default
+ * methods to push values into the element, which can be overridden by implementing classes.
+ */
+public interface RandomElement {
```
This is not a RandomJsonElement. This is a meta model element holding data used to generate random json.  
So I'd rather stick with the naming.

tschlat commented on connectors/citrus-openapi/src/main/java/org/citrusframework/openapi/random/RandomGenerator.java:
```diff
@@ -0,0 +1,60 @@
+ * <p>The class provides methods for determining if a generator can handle a given schema,
+ * based on the schema type, format, pattern, and enum constraints.
+ */
+public abstract class RandomGenerator {
```
I am willing to restructure this but let us postpone this to antoher PR.

tschlat commented on connectors/citrus-openapi/src/main/java/org/citrusframework/openapi/random/RandomModelBuilder.java:
```diff
@@ -0,0 +1,133 @@
+ * String json = builder.writeToJson();
+ * </pre>
+ */
+public class RandomModelBuilder {
```
I decided against this. The model which is build by the builder is a meta model that contains structured data. It is not a Json model by its pure nature. Instead, it can finally be converted to JSON using a specific writer. So in principal it is possible to use the generator for other formats by providing a specific converter.

tschlat commented on connectors/citrus-openapi/src/main/java/org/citrusframework/openapi/random/RandomEnumGenerator.java:
```diff
@@ -0,0 +1,26 @@
+            String enumValues = schema.enum_.stream()
+                    .map(value -> "'" + value + "'")
+                    .collect(joining(","));
+            randomContext.getRandomModelBuilder().appendSimpleQuoted("citrus:randomEnumValue(%s)".formatted(enumValues));
```
Agreed. I have reworked some of the Generator code already, incl. this issue in another PR. So I would like to postpone this, together with the previous issue.

tschlat commented on connectors/citrus-openapi/src/main/java/org/citrusframework/openapi/random/RandomModelWriter.java:
```diff
@@ -0,0 +1,112 @@
+ * Utility class for converting a {@link RandomModelBuilder} to its string representation.
+ * This class provides static methods to serialize the model built by {@link RandomModelBuilder}.
+ */
+final class RandomModelWriter {
```
I renamed it to RandomModelJsonWriter

tschlat commented on connectors/citrus-openapi/src/main/java/org/citrusframework/openapi/random/RandomNumberGenerator.java:
```diff
@@ -0,0 +1,152 @@
+public class RandomNumberGenerator extends RandomGenerator {
+
+    public static final BigDecimal THOUSAND = new BigDecimal(1000);
+    public static final BigDecimal HUNDRED = java.math.BigDecimal.valueOf(100);
```
Switched all to valueOf without qualifying the classname.

tschlat commented on connectors/citrus-openapi/src/main/java/org/citrusframework/openapi/random/RandomModelWriter.java:
```diff
@@ -0,0 +1,112 @@
+            Object value = entry.getValue();
+
+            if ("ARRAY".equals(key)) {
+                appendObject(builder, value);
```
Good catch. I have checked the code and this snippet can never be called. I think it is a relict. Actually the whole method can be removed as far as i can see.

tschlat commented:
> Thanks for your work! In general I am good with the PR.
> 
> I have opened a review PR #1332 to incorporate some improvements (e.g. missing license headers)
> 
> As a follow-up it would be good to think about what needs to be adjusted in the documentation regarding these changes made here.

I have adjusted the documentation in my last MR. So we can check there.

tschlat commented on connectors/citrus-openapi/src/test/java/org/citrusframework/openapi/OpenApiMessageTypeTest.java:
```diff
@@ -0,0 +1,18 @@
+package org.citrusframework.openapi;
```
This was added by Christop 🙇‍♂️

tschlat commented:
I have implemented all changes besides the ones left open regarding RandomObjectGenerator. I will provide fixes for them with the next mr.

tschlat commented:
Sorry for beeing so late on this. I had to fix several issues with respect to ReceiveMessageAction and reconcile validation context. 
I have implemented required changes and rebased to main.

bbortt commented on tools/jbang/dist/CitrusJBang.java:
```diff
@@ -18,10 +18,10 @@
+//DEPS org.citrusframework:citrus-bom:${citrus.jbang.version:4.7.0-SNAPSHOT}@pom
+//DEPS org.citrusframework:citrus-jbang:${citrus.jbang.version:4.7.0-SNAPSHOT}
+//DEPS org.citrusframework:citrus-agent:${citrus.jbang.version:4.7.0-SNAPSHOT}
+//DEPS org.citrusframework:citrus-agent-connector:${citrus.jbang.version:4.7.0-SNAPSHOT}
```
I wonder why this hasn't been touched by the release process @christophd ..?

bbortt reviewed (commented):
https://github.com/citrusframework/citrus/pull/1325/commits/2665fb209930e5fcfe81560ff6012b77f687d995 looks good to me. thanks!

bbortt commented:
thank you so much. I will provide the discussed fix for references shortly. feel free to proceed with part (MR) 4 in the meantime @tschlat 😉
